tfhe/src/integer/server_key/radix_parallel/scalar_bitwise_op.rs

```diff
@@ -0,0 +1,390 @@
+use rayon::prelude::*;
+
+impl ServerKey {
+    pub fn unchecked_scalar_bitand_parallelized<T>(
+        &self,
+        lhs: &RadixCiphertext,
+        rhs: T,
+    ) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitand_assign_parallelized(&mut result, rhs);
+        result
+    }
+
+    pub fn unchecked_scalar_bitand_assign_parallelized<T>(&self, lhs: &mut RadixCiphertext, rhs: T)
+    where
+        T: DecomposableInto<u8>,
+    {
+        let message_modulus = self.key.message_modulus.0;
+        assert!(message_modulus.is_power_of_two());
+
+        let clear_blocks = BlockDecomposer::with_early_stop_at_zero(rhs, message_modulus.ilog2())
+            .iter_as::<u8>()
+            .collect::<Vec<_>>();
+
+        lhs.blocks
+            .par_iter_mut()
+            .zip(clear_blocks.par_iter().copied())
+            .for_each(|(lhs_block, clear_block)| {
+                self.key
+                    .unchecked_scalar_bitand_assign(lhs_block, clear_block);
+            });
+
+        // Blocks beyond clear_blocks.len() should be 'bitanded'
+        // with '0', however, no matter the block value the result will be 0
+        let n = clear_blocks.len();
+        for block in &mut lhs.blocks[n..] {
+            self.key.create_trivial_assign(block, 0);
+        }
+    }
+
+    pub fn unchecked_scalar_bitor_parallelized<T>(
+        &self,
+        lhs: &RadixCiphertext,
+        rhs: T,
+    ) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitor_assign_parallelized(&mut result, rhs);
+        result
+    }
+
+    pub fn unchecked_scalar_bitor_assign_parallelized<T>(&self, lhs: &mut RadixCiphertext, rhs: T)
+    where
+        T: DecomposableInto<u8>,
+    {
+        let message_modulus = self.key.message_modulus.0;
+        assert!(message_modulus.is_power_of_two());
+
+        let clear_blocks = BlockDecomposer::with_early_stop_at_zero(rhs, message_modulus.ilog2())
+            .iter_as::<u8>()
+            .collect::<Vec<_>>();
+
+        lhs.blocks
+            .par_iter_mut()
+            .zip(clear_blocks.par_iter().copied())
+            .for_each(|(lhs_block, clear_block)| {
+                self.key
+                    .unchecked_scalar_bitor_assign(lhs_block, clear_block);
+            });
+
+        // Blocks beyond clear_blocks.len() should be 'ored'
+        // with '0', which means they keep their value
+    }
+    pub fn smart_scalar_bitand_parallelized<T>(
+        &self,
+        lhs: &mut RadixCiphertext,
+        rhs: T,
+    ) -> RadixCiphertext
+    where
+        T: DecomposableInto<u8>,
+    {
+        if !lhs.block_carries_are_empty() {
+            self.full_propagate(lhs);
+        }
+        let mut result = lhs.clone();
+        self.unchecked_scalar_bitand_assign_parallelized(&mut result, rhs);
+        result
```

Comment:
can we call the smart assign ?

Reply:
I don't think so, if we do 

```rust
    pub fn smart_scalar_bitand_parallelized<T>(
        &self,
        lhs: &mut RadixCiphertext,
        rhs: T,
    ) -> RadixCiphertext
    where
        T: DecomposableInto<u8>,
    {
        let mut result = lhs.clone();
        self.smar_scalar_bitand_assign_parallelized(&mut result, rhs);
        result
}
```

then the carry propagation happens on a copy of lhs, not lhs itself, negating the whole point of smart operation

Reply:
you are right, so needs to be "duplicated"